crypto/fipsmodule/rand/fork_detect.c

```diff
@@ -42,37 +42,102 @@ DEFINE_STATIC_MUTEX(g_fork_detect_lock)
 DEFINE_BSS_GET(volatile char *, g_fork_detect_addr)
 DEFINE_BSS_GET(uint64_t, g_fork_generation)
 DEFINE_BSS_GET(int, g_ignore_madv_wipeonfork)
+DEFINE_BSS_GET(int, g_ignore_pthread_atfork)
 
+static int init_fork_detect_madv_wipeonfork(void *addr, long page_size) {
+
+  // Some versions of qemu (up to at least 5.0.0-rc4, see linux-user/syscall.c)
+  // ignore |madvise| calls and just return zero (i.e. success). But we need to
+  // know whether MADV_WIPEONFORK actually took effect. Therefore try an invalid
+  // call to check that the implementation of |madvise| is actually rejecting
+  // unknown |advice| values.
+  if (madvise(addr, (size_t)page_size, -1) == 0 ||
+      madvise(addr, (size_t)page_size, MADV_WIPEONFORK) != 0) {
+    munmap(addr, (size_t)page_size);
+    return 0;
+  }
+
+  return 1;
+}
+
+static void pthread_atfork_on_fork(void) {
+
+  struct CRYPTO_STATIC_MUTEX *const lock = g_fork_detect_lock_bss_get();
+
+  // Aquire locks to be on the safe side. We want to avoid the checks in
+  // |CRYPTO_get_fork_generation| getting executed before setting the sentinel
+  // flag. The write lock prevents any other thread from owning any other type
+  // of lock.
+  CRYPTO_STATIC_MUTEX_lock_write(lock);
+  volatile char *const flag_ptr = *g_fork_detect_addr_bss_get();
+  *flag_ptr = 0;
```

Comment:
I'm probably missing something, but I don't see what happens when this flag is 0 and what sets it back to 1. I'm assuming that's the sentinel. So I was wondering about the handling that happens as in s2n: https://github.com/aws/s2n-tls/blob/main/utils/s2n_random.c#L127-L147

Reply:
That's right. I almost think that this PR is easier to review when looking at the entire raw file and not just the Git diff overview.

The sentinel is reset in `CRYPTO_get_fork_detection()`. When a fork has been detected one of the two detection methods flips the sentinel. This is detected in `CRYPTO_get_fork_detection()` that increments the cached generation number. It then flips back the sentinel to 1.

Detects fork and flips sentinel: https://github.com/awslabs/aws-lc/blob/main/crypto/fipsmodule/rand/fork_detect.c#L114-L118

Reply:
Thanks for the clarification.